src/domain/transaction/stellar/stellar_transaction.rs

```diff
@@ -171,6 +176,166 @@ where
+    async fn finalize_transaction_state(
+        &self,
+        tx_id: String,
+        new_status: TransactionStatus,
+        status_reason: Option<String>,
+        confirmed_at: Option<String>,
+    ) -> Result<TransactionRepoModel, TransactionError> {
+        let update_req = TransactionUpdateRequest {
+            status: Some(new_status),
+            status_reason,
+            confirmed_at,
+            ..Default::default()
+        };
+
+        let updated_tx = self
+            .transaction_repository()
+            .partial_update(tx_id, update_req)
+            .await?;
+
+        self.send_transaction_update_notification(&updated_tx)
+            .await?;
+        Ok(updated_tx)
+    }
+
+    /// Helper function to re-queue a transaction status check job.
+    async fn requeue_status_check(
+        &self,
+        tx: &TransactionRepoModel,
+    ) -> Result<(), TransactionError> {
+        self.job_producer()
+            .produce_check_transaction_status_job(
+                TransactionStatusCheck::new(tx.id.clone(), tx.relayer_id.clone()),
+                Some(STELLAR_DEFAULT_STATUS_RETRY_DELAY_SECONDS),
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// Parses the transaction hash from the network data and validates it.
+    /// Returns a `TransactionError::ValidationError` if the hash is missing, empty, or invalid.
+    fn parse_and_validate_hash(&self, tx: &TransactionRepoModel) -> Result<Hash, TransactionError> {
+        let stellar_network_data = tx.network_data.get_stellar_transaction_data()?;
+
+        let tx_hash_str = stellar_network_data.hash.as_deref().filter(|s| !s.is_empty()).ok_or_else(|| {
+            TransactionError::ValidationError(format!(
+                "Stellar transaction {} is missing or has an empty on-chain hash in network_data. Cannot check status.",
+                tx.id
+            ))
+        })?;
+
+        let stellar_hash: Hash = tx_hash_str.parse().map_err(|e: Error| {
+            TransactionError::UnexpectedError(format!(
+                "Failed to parse transaction hash '{}' for tx {}: {:?}. This hash may be corrupted or not a valid Stellar hash.",
+                tx_hash_str, tx.id, e
+            ))
+        })?;
+
+        Ok(stellar_hash)
+    }
+
+    /// Handles the logic when a Stellar transaction is confirmed successfully.
+    async fn handle_stellar_success(
+        &self,
+        tx: TransactionRepoModel,
+        _provider_response: soroban_rs::stellar_rpc_client::GetTransactionResponse, // May be needed later for more details
+    ) -> Result<TransactionRepoModel, TransactionError> {
+        let confirmed_tx = self
+            .finalize_transaction_state(
+                tx.id.clone(),
+                TransactionStatus::Confirmed,
+                None,
+                Some(Utc::now().to_rfc3339()),
+            )
+            .await?;
+
+        if let Some(oldest_pending) = self
```

Comment:
Do we need to add logic to compare the chain’s current sequence number with the next transaction’s sequence number to cover all cases, or is it safe to assume they’re always in order?

I can imagine a scenario where our job handler marks a transaction as mined/confirmed and immediately creates a new job to submit another pending transaction. At the same time, the API request handler may see no “in-flight” transactions and send a newly received one instead of waiting for the existing pending jobs to execute. This race could lead to out-of-order sequence numbers.

Also there is use case when one tx fails, what to do.

Reply:
Thanks for pointing this out, the logic wasn't ideal and there were still race conditions happening when i went back to test. I have refactored the gating mechanism completely. I have added new file `lane_gate` which implements a hashmap with a mutex to track which relayer transaction currently owns the submission lane. After each confirmed transaction we find the next tx, pass ownership to it and then enqueue it.